### PR TITLE
Ms hdf

### PR DESCRIPTION
**Pull request by martinspetlik**

### CODE REVIEW

File: src/hdf.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            self.base_structure(work_dir, step_range, job_dir, n_levels)
+        # Hdf file object with mode append
+        self._hdf_file_append = h5py.File(self._file_name, 'a')
+        # Hdf file object with read mode
+        self._hdf_file_reader = h5py.File(self._file_name, 'r')
+
+        # HDF absolute path to level groups {level_id: path in hdf file}
+        self._level_group_hdf_paths = {}
+
+    def base_structure(self, work_dir, step_range, job_dir, n_levels):
```
Comment (jbrezmorf) [inline src/hdf.py]: - init_header(), pass non file main attributes
- attributes for stored attributes

File: src/hdf.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+            # Set global attributes
+            writer.attrs['version'] = '1.0.1'
+            writer.attrs['work_dir'] = work_dir
+            writer.attrs['job_dir'] = job_dir
+            writer.attrs.create("n_levels", n_levels, dtype=np.int8)
+            # Create groups Levels and set step range attribute
+            writer.create_group("Levels")
+            writer["Levels"].attrs.create("step_range", step_range, dtype=np.float16)
+
+    def create_level_group(self, level_id):
```
Comment (jbrezmorf) [inline src/hdf.py]: rename: add_level_group
return: LevelGroup

read trough property level_groups[level_id]

File: src/hdf.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+            # Create group for each level named by level id (e.g. 0, 1, 2, ...)
+            self._hdf_file_append['Levels'].create_group(level_id)
+            # Set level id as group attribute
+            self._hdf_file_append[level_group_hdf_path].attrs.create("id", level_id, dtype=np.int8)
+        # @TODO: level step as attribute
+
+        return level_group_hdf_path
+
+    # @TODO: create property and setter for n_ops_estimate, or rather general method for getting and setting level group attributes
+    def set_n_ops_estimate(self, level_id, n_ops_estimate):
```
Comment (jbrezmorf) [inline src/hdf.py]: Pass as add_level parameter.

File: src/hdf.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    #     Create hdf file with default metadata
+    #     :param hdf_file:
+    #     :param attributes:
+    #     :return:
+    #     """
+    #     with h5py.File(self._file_name, 'w') as writer:
+    #         for name, value in attributes.items():
+    #             writer.attrs[name] = value
+
+    def save_scheduled(self, level_id, scheduled_samples):
```
Comment (jbrezmorf) [inline src/hdf.py]: - rename: append_scheduled
- method of LeveLGroup

File: src/hdf.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+            HDF5.DATASET_NAMES['scheduled'],
+            shape=(len(scheduled_samples),),
+            maxshape=(None,),
+            dtype=HDF5.SCHEDULED_DTYPE,
+            chunks=True)
+
+        # Set values to dataset
+        scheduled_dset[()] = scheduled_samples
+
+    def change_scheduled_sample_status(self, level_id, sample_id, status):
```
Comment (jbrezmorf) [inline src/hdf.py]: Remove/comment.

File: src/hdf.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        """
+        scheduled_dataset = '/'.join([self._level_group_hdf_paths[level_id], HDF5.DATASET_NAMES['scheduled']])
+        if scheduled_dataset not in self._hdf_file_append:
+            raise Exception("Dataset is not in HDF5 file")
+        else:
+            sample_row = self._hdf_file_append[scheduled_dataset][sample_id]
+            sample_row[-1] = status
+            self._hdf_file_append[scheduled_dataset][sample_id] = sample_row
+
+    def save_collected(self, level_id, collected_samples, rewrite=False):
```
Comment (jbrezmorf) [inline src/hdf.py]: - append_collected
- replace rewrite by: save_failed

File: src/hdf.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+            dataset = self._hdf_file_append[self._level_group_hdf_paths[level_id]].create_dataset(
+                HDF5.COLLECTED_ATTRS[attr_name]['name'],
+                shape=shape,
+                maxshape=maxshape,
+                dtype=HDF5.COLLECTED_ATTRS[attr_name]['dtype'],
+                chunks=True)
+            # Set data
+            dataset[()] = data
+
+    def sample_attributes_pairs(self, fine_coarse_samples):
```
Comment (jbrezmorf) [inline src/hdf.py]: private _

File: src/hdf.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+        """
+        if dataset_hdf_path not in self._hdf_file_reader:
+            raise Exception("Dataset doesn't exist")
+        # Resize dataset
+        self._hdf_file_append[dataset_hdf_path].resize(self._hdf_file_append[dataset_hdf_path].shape[0] +
+                                                       len(values), axis=0)
+        # Append new values to the end of dataset
+        self._hdf_file_append[dataset_hdf_path][-len(values):] = values
+
+    def read_scheduled(self, level_id):
```
Comment (jbrezmorf) [inline src/hdf.py]: - method of LevelGroup
- rename to  scheduled()
usage:
```
hdf5.level_groups[level_id].scheduled()
```

File: src/hdf.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+            return
+        else:
+            # Scheduled dataset
+            dataset = self._hdf_file_reader[dataset_hdf_path]
+
+            # Return dataset items
+            for dset_value in dataset:
+                yield dset_value
+
+    def read_collected(self, level_id):
```
Comment (jbrezmorf) [inline src/hdf.py]: similar to scheduled

File: setup.py [hunk 10]
```diff
@@ -30,7 +30,7 @@ def read(*names, **kwargs):
 
 
 setup(
-    # Project (package) name
+    # Project (job_id) name
```
Comment (jbrezmorf) [inline setup.py]: Rename back.

File: src/mlmc/hdf/hdf.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
+import h5py
+import numpy as np
+import os
+from mlmc.hdf.level_group import LevelGroup
+
+
+class HDF5:
+    # @TODO: comment hdf structure
```
Comment (jbrezmorf) [inline src/mlmc/hdf/hdf.py]: TODO

File: src/mlmc/hdf/level_group.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+    # {attribute name in class Sample : {name: dataset name, maxshape: dataset maxshape, dtype: dataset dtype}}
+    COLLECTED_ATTRS = {"sample_id": {'name': 'collected_ids', 'maxshape': (None, 1), 'dtype': np.int16},
+                       "result": {'name': 'collected_values', 'maxshape': (None, 2, None), 'dtype': np.float64},
+                       "time": {'name': 'collected_times', 'maxshape': (None, 2, None), 'dtype': np.float64}}
+
+    DATASET_NAMES = {"scheduled": "scheduled",
+                     "collected_ids": "collected_ids"}
+
+    def __init__(self, hdf_group, level_id, job_dir):
+        # HDF5 level group
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Improve doc. Doc string for function. Doc for attributes, precise reference to the class of hdf_group.

File: src/mlmc/hdf/level_group.py [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+            # Set data
+            dataset[()] = data
+
+    def _sample_attributes_pairs(self, fine_coarse_samples):
+        """
+        Merge fine sample and coarse sample collected values to one dictionary
+        :param fine_coarse_samples: list of tuples; [(Sample(), Sample()), ...]
+        :return: zip
+        """
+        fine_coarse_data = np.array([(f_sample.collected_data_array(LevelGroup.COLLECTED_ATTRS),
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Better readibility.

File: src/mlmc/hdf/level_group.py [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+                dtype=LevelGroup.COLLECTED_ATTRS[attr_name]['dtype'],
+                chunks=True)
+            # Set data
+            dataset[()] = data
+
+    def _sample_attributes_pairs(self, fine_coarse_samples):
+        """
+        Merge fine sample and coarse sample collected values to one dictionary
+        :param fine_coarse_samples: list of tuples; [(Sample(), Sample()), ...]
+        :return: zip
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Describe return value.

File: src/mlmc/hdf/level_group.py [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+                       'formats': (SAMPLE_DTYPE, SAMPLE_DTYPE)}
+
+    # Data that is collected, only this data can by saved to datasets
+    # @TODO change docstring
+    # {attribute name in class Sample : {name: dataset name, maxshape: dataset maxshape, dtype: dataset dtype}}
+    COLLECTED_ATTRS = {"sample_id": {'name': 'collected_ids', 'maxshape': (None, 1), 'dtype': np.int16},
+                       "result": {'name': 'collected_values', 'maxshape': (None, 2, None), 'dtype': np.float64},
+                       "time": {'name': 'collected_times', 'maxshape': (None, 2, None), 'dtype': np.float64}}
+
+    DATASET_NAMES = {"scheduled": "scheduled",
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Use these names directly, but only at two places:
- creation of the data set
- property (e.g. LevelGroup.collected_ids_ds() that returns particular dataset)

File: src/mlmc/hdf/level_group.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+        if LevelGroup.DATASET_NAMES['scheduled'] not in self._level_group:
+            self._create_scheduled_dataset(sim_hdf_form)
+        # Dataset scheduled already exists
+        else:
+            self._append_dataset(LevelGroup.DATASET_NAMES['scheduled'], sim_hdf_form)
+
+        # Save to jobs to datasets
+        self._append_jobs(jobs)
+
+    def _create_scheduled_dataset(self, scheduled_samples):
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: This should only create the dataset, adding samples through append_samples even at first append.

File: src/mlmc/hdf/level_group.py [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+        # Go through scheduled samples and format them
+        for sample_id, (fine_sample, coarse_sample) in scheduled_samples.items():
+            sim_hdf_form.append((fine_sample.scheduled_data(), coarse_sample.scheduled_data()))
+            # Append sample id to job id
+            jobs.setdefault(fine_sample.job_id, set()).add(fine_sample.sample_id)
+            if int(self.level_id) != 0:
+                jobs.setdefault(coarse_sample.job_id, set()).add(coarse_sample.sample_id)
+
+        # Create dataset scheduled
+        if LevelGroup.DATASET_NAMES['scheduled'] not in self._level_group:
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Why not creating all datasets in constructor (or method called from constructor)?

File: src/mlmc/hdf/level_group.py [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+        # Set values to dataset
+        scheduled_dset[()] = scheduled_samples
+
+    def _append_jobs(self, jobs):
+        """
+        Save jobs to datasets
+        :param jobs: dict, {job_id: [sample_id,...], ...}
+        :return: None
+        """
+        if 'Jobs' not in self._level_group:
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Create in constructor?

File: src/mlmc/hdf/level_group.py [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+            self._level_group.create_group('Jobs')
+
+        for job_name, job_samples in jobs.items():
+            job_dataset_path = '/'.join(['Jobs', job_name])
+
+            if job_dataset_path not in self._level_group:
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Make a method for safe adding of the dataset, pass all parameters through.

File: src/mlmc/hdf/level_group.py [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+                self._level_group.create_dataset(job_dataset_path, data=list(job_samples), maxshape=(None,),
+                                                     chunks=True)
+            else:
+                self._append_dataset('/'.join([self._level_group.name, job_dataset_path]), list(job_samples))
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: better:
dset = create_dataset(...)
self._append_dataset(dset, job_dataset_path) # also simplifies this method

File: src/mlmc/hdf/level_group.py [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+
+    def append_collected(self, collected_samples):
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: simplify.

File: src/mlmc/hdf/level_group.py [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+                data=list(failed_samples),
+                maxshape=(None,),
+                chunks=True)
+            # Set data
+        else:
+            dataset = self._level_group['failed_ids']
+            dataset.resize((len(failed_samples),))
+            dataset[()] = list(failed_samples)
+
+    def _append_dataset(self, dataset_name, values):
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Pass directly dataset to append.

File: src/mlmc/hdf/level_group.py [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+        if 'failed_ids' not in self._level_group:
+            # Create resizable dataset
+            self._level_group.create_dataset(
+                'failed_ids',
+                data=list(failed_samples),
+                maxshape=(None,),
+                chunks=True)
+            # Set data
+        else:
+            dataset = self._level_group['failed_ids']
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: Can we use _append_dataset?

File: src/mlmc/hdf/level_group.py [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+        # Number of collected samples
+        num_collected_samples = self._level_group[LevelGroup.DATASET_NAMES['collected_ids']].len()
+
+        # Collected data as dictionary according to lookup table COLLECTED_ATTRS
+        # dictionary format -> {COLLECTED_ATTRS key: corresponding dataset values generator, ...}
+        collected_data = {sample_attr_name: self._dataset_values('/'.join([self._level_group.name, dset_params['name']]))
+                          for sample_attr_name, dset_params in LevelGroup.COLLECTED_ATTRS.items()}
+
+        # For each item create two Sample() instances - fine sample and coarse sample
+        for _ in range(num_collected_samples):
```
Comment (jbrezmorf) [inline src/mlmc/hdf/level_group.py]: This seems to pass through the dataset one time for every attribute, bad idea.
Just iterate through dataset and create two instances of Sample for every row. Can be done with usage of COLLECTED_ATTRS, that 's fine.

File: src/mlmc/logger.py [hunk 25]
```diff
@@ -3,71 +3,143 @@
 import json
```
Comment (jbrezmorf) [inline src/mlmc/logger.py]: Try to get rid of the logger.

File: src/mlmc/hdf/level_group.py (hunk 24 above)
Comment (martinspetlik) [reply src/mlmc/hdf/level_group.py]: It is loop through range of number of samples, for each iteration there is loop through collected attributes ({attr_name: dataset generator, ...}). Sample() instance is created from next() value of datasets. So in fact there are more datasets and generator should provide next element of each one.

Comment (jbrezmorf) [reply src/mlmc/hdf/level_group.py]: I see. OK, concept is fine. Temporary creation of the dictionaries for every sample may be slow, but will see later.

**Pull request merged**